fili-core/src/main/java/com/yahoo/bard/webservice/table/availability/MetricUnionAvailability.java

```diff
@@ -0,0 +1,190 @@
+                                        )
+                        )
+                );
+
+        // validate metric uniqueness such that
+        // each table's underlying datasource schema don't have repeated metric column
+        Map<MetricColumn, Set<Availability>> duplicates = getDuplicateValue(metricAvailability);
+        if (!duplicates.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format(
+                            "Metric columns are not unique - {}",
+                            duplicates
+                    )
+            );
+        }
+
+        dataSourceNames = physicalTables.stream()
+                .map(PhysicalTable::getTableName)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public Set<TableName> getDataSourceNames() {
+        return dataSourceNames;
+    }
+
+    @Override
+    public Map<Column, List<Interval>> getAllAvailableIntervals() {
+        return metricAvailability.keySet().stream()
+                .map(availability -> availability.getAllAvailableIntervals().entrySet())
+                .flatMap(Set::stream)
+                .filter(entry -> metricColumns.contains(entry.getKey()))
+                .collect(
+                        Collectors.toMap(
+                                entry -> entry.getKey(),
+                                entry -> entry.getValue(),
+                                (value1, value2) -> {
+                                    List<Interval> combined = new ArrayList<>();
+                                    combined.addAll(value1);
+                                    combined.addAll(value2);
+                                    return combined;
+                                }
+                        )
+                );
+    }
+
+    @Override
+    public SimplifiedIntervalList getAvailableIntervals(DataSourceConstraint constraints) {
+        return new SimplifiedIntervalList(
+                constructSubConstraint(constraints).entrySet().stream()
+                        .map(entry -> entry.getKey().getAvailableIntervals(entry.getValue()))
+                        .map(i -> (Set<Interval>) new HashSet<>(i))
+                        .reduce(null, IntervalUtils::getOverlappingSubintervals)
+        );
+    }
+
+    /**
+     * Returns duplicate values of <tt>MetricColumn</tt>'s in a map of <tt>Availability</tt>
+     * to a set of <tt>MetricColumn</tt>'s contained in that <tt>Availability</tt>.
+     * <p>
+     * The return value is a map of duplicate <tt>MetricColumn</tt> to all <tt>Availability</tt>'s that contains
+     * this <tt>MetricColumn</tt>
+     *
+     * @param availabilitySetMap  A map from <tt>Availability</tt> to set of <tt>MetricColumn</tt> contained in that
+     * <tt>Availability</tt>
+     *
+     * @return duplicate values of <tt>MetricColumn</tt>'s
+     */
+    private static Map<MetricColumn, Set<Availability>> getDuplicateValue(
```

Comment:
I've no problem with logging more stuff, but we should make sure that we're doing it at the right point, and keeping things as encapsulated and clear as we can. This class in general seems to be passing around a lot of Maps with signatures that are not particularly clear, so I want to make sure that we're letting the code be simple and easy to understand.